### Import dependencies

In [1]:
import pandas as pd
import os

## Deliverable 1: Collect the Data

Import the data from the new_full_student_data.csv file into a DataFrame named student_df by using the Pandas read_csv function and the os module.

Confirm that Pandas correctly imported the data by using the head function


In [4]:
# Import the data
student_data = os.path.join('/Users/nikkichappelle/resource/new_full_student_data.csv')
student_df = pd.read_csv(student_data)

In [5]:
# Verify that the data was properly imported by using the head function
student_df.head()

student_id      student_name grade           school_name  reading_score  \
0   103880842     Travis Martin   9th  Sullivan High School           59.0   
1    45069750     Michael Brown   9th     Dixon High School           94.7   
2    45024902   Gabriela Lucero   9th    Wagner High School           89.0   
3    62582498  Susan Richardson   9th     Silva High School           69.7   
4    16437227      Sherry Davis  11th    Bowers High School            NaN   

   math_score school_type  school_budget  
0        88.2      Public         961125  
1        73.5     Charter         870334  
2        70.4      Public         846745  
3        80.3      Public         991918  
4        27.5      Public         848324

## Deliverable 2: Prepare the Data

    
1. In the student DataFrame, check for rows that have NaN (or missing) values, and remove those rows,

2. In the student DataFrame, check for duplicate rows, and remove them.

3. Check the data types of the columns by using the dtypes property, as the following image shows:

4. In the grade column, remove the "th" suffix from every value by using str and replace, as the following image shows:

5. Change the "grade" column to the int type, and then verify the column types, as the following image shows:


In [6]:
# Check for rows that have NaN
student_df.isna().sum()

student_id          0
student_name        0
grade               0
school_name         0
reading_score    1968
math_score        982
school_type         0
school_budget       0
dtype: int64

In [7]:
# Removes rows that have NaN
student_df = student_df.dropna()
student_df.isna().sum()

student_id       0
student_name     0
grade            0
school_name      0
reading_score    0
math_score       0
school_type      0
school_budget    0
dtype: int64

In [8]:
# Check for duplicated rows
student_df.duplicated().sum()

1836

In [9]:
# Drop duplicated rows
student_df = student_df.drop_duplicates()  
student_df.duplicated().sum()

0

In [10]:
# Check data types
student_df.dtypes

student_id         int64
student_name      object
grade             object
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

In [11]:
# Remove the "th" suffix from every value by using str and replace
prefixes_suffixes = ["th"]
for word in prefixes_suffixes:
    student_df["grade"] = student_df["grade"].str.replace(word,"")
student_df["grade"]

0         9
1         9
2         9
3         9
5         9
         ..
19508    10
19509    12
19511    11
19512    11
19513    12
Name: grade, Length: 14831, dtype: object

In [12]:
# Change the "grade" column to the int type, and then verify the column types
student_df.loc[:, "grade"] = student_df.loc[:, "grade"].astype("int")
student_df.dtypes

student_id         int64
student_name      object
grade              int64
school_name       object
reading_score    float64
math_score       float64
school_type       object
school_budget      int64
dtype: object

## Deliverable 3: Summarize the Data

Describe the data using summary statistics on the data as a whole and on individual columns.

1. Generate the summary statistics for each DataFrame by using the `describe` function.

2. Display the mean math score using the `mean` function. 

2. Store the minimum reading score as `min_reading_score`.

In [13]:
# Generate the summary statistics for each DataFrame by using the `describe` function.
student_df.describe()

student_id         grade  reading_score    math_score  school_budget
count  1.483100e+04  14831.000000   14831.000000  14831.000000   14831.000000
mean   6.975296e+07     10.355539      72.357865     64.675733  893742.749107
std    3.452909e+07      1.097728      15.224590     15.844093   53938.066467
min    1.000906e+07      9.000000      10.500000      3.700000  817615.000000
25%    3.984433e+07      9.000000      62.200000     54.500000  846745.000000
50%    6.965978e+07     10.000000      73.800000     65.300000  893368.000000
75%    9.927449e+07     11.000000      84.000000     76.000000  956438.000000
max    1.299997e+08     12.000000     100.000000    100.000000  991918.000000

In [14]:
# Display the mean math score using the mean function
student_df["math_score"].mean()

64.67573326141189

In [15]:
# Store the minimum reading score as min_reading_score
min_reading_score  = student_df["reading_score"].min()
min_reading_score

10.5

## Deliverable 4: Drill Down into the Data


1. Display the grade column by using loc

2. Display the first three rows of Columns 3, 4, and 5 by using iloc, as the following image shows

3. Select the rows for Grade 9, and display their summary statistics by using loc and describe

4. Store the row with the minimum overall reading score in min_reading_row by using loc and the min_reading_score variable from Deliverable 3

5. Select all the reading scores from the 10th graders at Dixon High School by using loc with conditionals

6. Find the mean reading score for all the students in Grades 11 and 12 combined by using conditional statements and loc or iloc

In [16]:
# Use loc to display the grade column
student_df.loc[:,["grade"]]

grade
0          9
1          9
2          9
3          9
5          9
...      ...
19508     10
19509     12
19511     11
19512     11
19513     12

[14831 rows x 1 columns]

In [17]:
# Display the first three rows of Columns 3, 4, and 5 by using iloc
student_df.iloc[[0,1,2], [3,4,5]]

school_name  reading_score  math_score
0  Sullivan High School           59.0        88.2
1     Dixon High School           94.7        73.5
2    Wagner High School           89.0        70.4

In [18]:
# Select the rows for grade nine and display their summary statistics using `loc` and `describe`.
grade9_schools_df = student_df.loc[student_df["grade"] == 9]
grade9_schools_df.describe()

student_id   grade  reading_score   math_score  school_budget
count  4.132000e+03  4132.0    4132.000000  4132.000000    4132.000000
mean   6.979441e+07     9.0      69.236713    66.585624  898692.606002
std    3.470565e+07     0.0      15.277354    16.661533   54891.596611
min    1.000906e+07     9.0      17.900000     5.300000  817615.000000
25%    3.953848e+07     9.0      59.000000    56.000000  846745.000000
50%    6.984037e+07     9.0      70.050000    67.800000  893368.000000
75%    9.939504e+07     9.0      80.500000    78.500000  957299.000000
max    1.299997e+08     9.0      99.900000   100.000000  991918.000000

In [19]:
# Store the row with the minimum overall reading score as `min_reading_row`
# using `loc` and the `min_reading_score` found in Deliverable 3.
min_reading_row = student_df.loc[student_df["reading_score"] == min_reading_score]
min_reading_row

student_id    student_name  grade        school_name  reading_score  \
3706    81758630  Matthew Thomas     10  Dixon High School           10.5   

      math_score school_type  school_budget  
3706        58.4     Charter         870334

In [29]:
# Use loc with conditionals to select all reading scores from 10th graders at Dixon High School.
school = student_df.loc[student_df["school_name"] == 'Dixon High School']
student_df.loc[student_df["grade"] == 10]
school.loc[:, ["school_name", "reading_score"]]


school_name  reading_score
1      Dixon High School           94.7
32     Dixon High School           90.3
45     Dixon High School           71.1
58     Dixon High School           83.5
60     Dixon High School           59.5
...                  ...            ...
19401  Dixon High School           62.1
19435  Dixon High School           61.0
19445  Dixon High School           43.9
19466  Dixon High School           95.2
19477  Dixon High School           69.8

[1583 rows x 2 columns]

In [75]:
# Find the mean reading score for all students in grades 11 and 12 combined.
Mean_reading_12 = student_df.loc[student_df ["grade"] == 12, "reading_score"].mean()
Mean_reading_11 = student_df.loc[student_df ["grade"] == 11, "reading_score"].mean()
(Mean_reading_12 + Mean_reading_11)/2

74.8617601794854

## Deliverable 5: Make Comparisons Between District and Charter Schools


1. Display the average budget for each school type by using the groupby and mean functions.

2. Find the total number of students at each school, and sort those numbers from largest to smallest by using the groupby, count, and sort_values functions

3. Find the average math score by grade for each school type by using the groupby and mean functions, as the following image shows

In [42]:
# Display the average budget for each school type by using the groupby and mean functions.
school_budget_mean = student_df.groupby(["school_type"]).mean()["school_budget"]
school_budget_mean


school_type
Charter    872625.656236
Public     911195.558251
Name: school_budget, dtype: float64

In [47]:
# Find the total number of students at each school, and sort those numbers from largest to smallest by using the groupby, count, and sort_values functions
min_scores_by_school = student_df.groupby(by='student_id').count()
min_scores_by_school.loc[:, ["school_name", "student_id"]].sort_values("student_id")


KeyError: "['student_id'] not in index"

In [43]:
# Find the average math score by grade for each school type by using the groupby and mean functions, as the following image shows
avg_by_school = student_df.groupby(['school_type', 'grade']).mean()
avg_by_school.loc[:, ["reading_score", "math_score"]]

reading_score  math_score
school_type grade                           
Charter     9          68.676790   70.077874
            10         69.628822   66.443206
            11         80.596379   68.024735
            12         70.475421   60.212121
Public      9          69.687981   63.771066
            10         73.160461   63.764121
            11         73.486448   59.314337
            12         73.342380   63.568319

# Deliverable 6: Summarize Your Findings
In the cell below, write a few sentences to describe any discoveries you made while performing your analysis along with any additional analysis you believe would be worthwhile.

The math mean score is below:
student_df["math_score"].mean() = 64.67

I think this shows that some of the budget needs to go into help improving math scores since overall the score is on the lower side.  

Also it looks like public schools have a higher budget then charter schools. 
school_type
Charter    872625.656236
Public     911195.558251

However, when looking at average reading and math scores for the two school types, both scools have scores on the lower side, especially eith ninth grade readin and llth and 12th grade math scores. 

Additionally, I think there needs to be an anlaysis that groups by schools and lists all average reading and math scores per school and we can see which schools themselves have lower scores and use some of the budget to help improves scores at these specific schools. 

